scripts/download-binaries.js

```diff
@@ -0,0 +1,152 @@
+  return new Promise((resolve) =>
+    fs.exists('manifest.json', (exists) => {
+      if (exists) {
+        // get it from manifest file
+        const manifestJson = require('./../manifest.json');
+        resolve(manifestJson);
+      } else {
+        // ask git
+        exec("git log --pretty=format:'%H' -n 1 .", (error, stdout, stderr) => {
+          const gitsha = stdout;
+          resolve({
+            tag: undefined,
+            gitsha,
+          })
+        });
+      }
+    })
+  );
+}
+
+/**
+ * Convenience function for downloading prebuilds binaries
+ *
+ * @param {*} platform Operating system platform (linux, darwin or win32)
+ * @param {*} arch Architecture of prebuild binaries (ex: x64 or ia32)
+ * @param {*} abiName Node abi name (ex: 57, 64, 67 etc)
+ * @param {*} destFile Name of the file where the downloaded binary will be saved
+ * @param {*} gitsha Git sha for getting specific binaries based on which git revision
+ * they were built from
+ * @returns Promise that resolves when download succeeds and rejects when download fails
+ */
+function downloadBinary(platform, arch, abiName, destFile, gitsha) {
+  let binaryName;
+  if (gitsha) {
+    binaryName = [[abiName, pkgJson.version, platform, arch, gitsha].join('-'), 'node'].join('.');
+  } else {
+    binaryName = [[abiName, pkgJson.version, platform, arch].join('-'), 'node'].join('.');
+  }
+  const file = fs.createWriteStream(destFile);
+
+  return new Promise(function (resolve, reject) {
+    const url = [pkgJson.binary.host, binaryName].join('');
+    console.log('Downloading ', url);
+    const cb = (response) => {
+      const statusCode = response.statusCode;
+      if (statusCode === 200) {
+        response.pipe(file);
+        file.on('finish', function () {
+          resolve();
+        });
+        file.on('error', function (e) {
+          reject(e);
+        });
+      } else {
+        reject([url, statusCode]);
+      }
+    }
+
+    https.get(url, cb).on('error', function (e) {
+      console.log(`Got error: ${e.message}`);
+      console.log(`Retrying: ${e.message}`);
+      https.get(url, cb).on('error', function (e) {
+        console.log(`Got error on retry attempt: ${e.message}`);
+        reject(e);
+      });
+    });
+  });
+}
+
+const prebuildsDir = path.join(__dirname, '../', 'prebuilds');
+
+if (!fs.existsSync(prebuildsDir)) {
+  fs.mkdirSync(prebuildsDir);
+}
+
+readManifestData()
+.then((manifest) => {
+  const promises = [];
+  const binaryMatrix = {
+    darwin: ['x64'],
+    linux: ['x64'],
+    win32: ['x64', 'ia32'],
+  };
+
+  // If the tag is set in the manifest, do not fetch gitsha binaries, but tag binaries
+  const gitsha = (manifest.tag && manifest.tag !== '') ? undefined : manifest.gitsha;
+  Object.keys(binaryMatrix).forEach((platform) => {
+    binaryMatrix[platform].forEach((arch) => {
+      // Create platform-arch directory to host binaries
+      var binaryDir = path.join(prebuildsDir, [platform, arch].join('-'));
```

Comment:
Don't you get a double dash here when you don't have sha, when you actually don't want a trailing -.

Reply:
`gitsha` is not references in this line. I am sending it to `downloadBinary` which builds the file name based on whether the gitsha is provided or not!